Heartbeat autodiscover

andrewvc opened the pull request:
This is a complete PR for heartbeat automatic reload aside from tests. Once those are added it will be ready for a true review.

This shares the adapter used by metricbeat, and moves that to libbeat to keep things DRY.

exekias commented on heartbeat/heartbeat.reference.yml:
```diff
@@ -218,6 +218,21 @@ heartbeat.monitors:
+#          docker.container.image: redis
+#      config:
+#      - type: tcp
+#        hosts: ["localhost:${data.port}"]
```
Probably `"${data.host}:${data.port}"` is better?

exekias commented on libbeat/autodiscover/factoryadapter.go:
```diff
@@ -26,20 +26,20 @@ import (
 
-// AutodiscoverAdapter for Metricbeat modules
-type AutodiscoverAdapter struct {
+// FactoryAdapter for Metricbeat modules
```
you will need to update the docs here (Metricbeat references), probably explain a little bit how this takes a factory and give you an autodiscover adapter

exekias commented on heartbeat/beater/heartbeat.go:
```diff
@@ -84,16 +88,29 @@ func (bt *Heartbeat) Run(b *beat.Beat) error {
 	defer bt.scheduler.Stop()
+	defer func() {
+
+	}()
```
spurious defer

andrewvc replied:
Oops, good catch.

andrewvc commented:
Thanks for the review @exekias ! I'll ping you once I've had the chance to add tests.

ruflin commented on heartbeat/beater/heartbeat.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/elastic/beats/libbeat/autodiscover"
```
import mess :-)

ruflin commented on libbeat/autodiscover/factoryadapter.go:
```diff
@@ -26,20 +26,20 @@ import (
 
-// AutodiscoverAdapter for Metricbeat modules
-type AutodiscoverAdapter struct {
+// FactoryAdapter is an adapter that works with any cfgfile.RunnerFactory.
```
Good to see we generalised it.

andrewvc commented:
Tests added & rebased off master. Will move out of WIP once CI passes

andrewvc commented:
I believe the test failures are unrelated, but there's one for metricbeat autodiscovery, so I'll run the suite again.

andrewvc commented:
jenkins, retest this please

ruflin commented:
On travis the metricbeat autodiscovery fails which I assume is related.

andrewvc commented:
@ruflin Yep. I've been spending today tracking it down, the metricbeat changes should have been a noop, but clearly that is not the case.

Narrowing down the cause, at the moment it's looking like somehow the hint builder is not being registered somehow.

andrewvc commented:
OK, just pushed d74b291 which I believe should fix it.

andrewvc commented:
@ruflin this is basically done.

The lack of a docker-compose + Dockerfile was breaking this (and #8475 ). I opened #8491 to fix this as a singular PR, so that should be a blocker for this.

I just rebased this onto that PR.

ruflin commented on heartbeat/docker-compose.yml:
```diff
@@ -0,0 +1,41 @@
+    environment:
+    - REDIS_HOST=redis
+    - REDIS_PORT=6379
+    - KIBANA_HOST=kibana
```
Remove kibana

andrewvc replied:
Yeah, this are complicated now because they're split across this and #8491. This is already fixed there, let's close that one out and I'll rebase here.

ruflin commented on heartbeat/tests/system/test_autodiscovery.py:
```diff
@@ -0,0 +1,63 @@
+        )
+
+        proc = self.start_beat()
+        docker_client.images.pull('redis:3.2.4')
```
If we set TEST_ENVIRONMENT to true, these containers should already be running.

There are currently two models of testing against containers we have:

* Start everything at the beginning: Libbeat and others
* Start on request and stop afterwards: Metricbeat modules

We seem to mix the two here. For simplicity I would start with setting TEST_ENVIRONMENT to true and not handle container start / stop in here.

ruflin commented on heartbeat/docker-compose.yml:
```diff
@@ -0,0 +1,41 @@
+      kibana:        { condition: service_healthy }
+      redis:         { condition: service_healthy }
+
+  elasticsearch:
```
As the output in the system tests below is not Elasticsearch, do we need elasticsearch?

andrewvc replied:
Yep, we can get rid of that as well. should happen in the #8491 merge.

andrewvc commented:
@ruflin let's put this on pause till https://github.com/elastic/beats/pull/8491 is merged, since there's overlap on the docker stuff.

andrewvc commented:
Rebased off current master with #8491 merged in

ruflin commented on heartbeat/tests/system/test_autodiscovery.py:
```diff
@@ -0,0 +1,61 @@
+        import docker
+        docker_client = docker.from_env()
+
+        self.render_config_template(
```
For later: As we have ES running anyways, we could add some additional tests for http.

andrewvc replied:
Discussed this on slack with @ruflin , agreed that adding a second test was not appropriate due to the fact that we're testing autodiscovery, not reloading.

andrewvc commented:
@ruflin , it'd be nice to backport to 6.x I think.

I made the test not pin to a specific redis version, and added a reference yaml entry.

Also added a changelog entry.

andrewvc commented:
Reverted the version change, that's needed

andrewvc commented:
In 9624a5d I've added the shared docs for autodiscover. I don't see any docs beyond that for the other beats.

In the future we should probably add some heartbeat focused examples.

andrewvc commented:
Will do docs in a separate PR, didn't realize I needed more than just the reference to the shared autodiscover doc.

andrewvc commented:
Hmm, I borked the tests with some finagling of the AD tests. Not sure exactly how, but I think it's the looser condition I'm trying

andrewvc commented:
jenkins, retest this please

ruflin commented on heartbeat/tests/system/config/heartbeat.yml.j2:
```diff
@@ -1,3 +1,5 @@
+logging.level: debug
```
This should not have any impact as `-e` is used to capture the logs from stdout.

ruflin replied:
@andrewvc I would suggest to remove the line but don't want to break your tests (it shouldn't  ;-) ). I worry that this might cause confusion in the future.

ruflin commented on heartbeat/tests/system/test_autodiscovery.py:
```diff
@@ -0,0 +1,62 @@
+        self.wait_until(lambda: self.log_contains(re.compile('autodiscover.+Got a start event:', re.I)))
+
+        self.wait_until(lambda: self.output_count(lambda x: x >= 1))
+        container.stop()
```
This seems the line it fails. As redis is started outside, no need to stop the container here. I assume varialbe is not initialised?

andrewvc commented:
Fixed some bugs, but this exposed a flimsy way of enumerating over docker hosts in the autodiscover tests in python in the tests. This needs to apply the same filter there.

andrewvc commented:
hmmm, these pass locally but not remotely... will take another look tomorrow
